src/Pos/Generator/Block/Payload.hs

```diff
@@ -0,0 +1,199 @@
+        -- utxo contains only related to these secret keys only.
+        let resolveSecret stId =
+                fromMaybe (error $ "can't find stakeholderId " <>
+                           pretty stId <> " in secrets map")
+                          (HM.lookup stId secrets)
+        let utxoAddresses = map (makePubKeyAddress . toPublic) $ HM.elems secrets
+
+        ----- INPUTS
+
+        inputsN <- getRandomR (1, min 5 utxoSize)
+        inputsIxs <- selectDistinct inputsN (0, utxoSize - 1)
+        -- It's alright to use unsafeIndex because length of
+        -- gtdUtxoKeys must match utxo size and inputsIxs is selected
+        -- prior to length limitation.
+        txIns <- forM inputsIxs $ \i -> uses gtdUtxoKeys (`V.unsafeIndex` i)
+        inputsResolved <- forM txIns $ \txIn ->
+            -- we're selecting from utxo by 'gtdUtxoKeys'. Inability to resolve
+            -- txin means that 'GenTxData' is malformed.
+            toaOut . fromMaybe (error "genTxPayload: inputsSum can't happen") <$> utxoGet txIn
+        let (inputsSum :: Integer) = sumCoins $ map txOutValue inputsResolved
+
+        ----- OUTPUTS
+
+        outputsMaxN <- fromIntegral . max 1 <$> lift (view tgpMaxOutputs)
+        (outputsN :: Int) <- fromIntegral <$> getRandomR (1, min outputsMaxN inputsSum)
+        outputsIxs <- selectDistinct outputsN (0, max outputsN (length utxoAddresses - 1))
+        let outputAddrs = map ((cycle utxoAddresses) !!) outputsIxs
+        -- We operate small coins values so any input sum mush be less
+        -- than coin maxbound.
+        coins <- splitCoins outputsN (unsafeIntegerToCoin inputsSum)
+        let txOuts = NE.fromList $ zipWith TxOut outputAddrs coins
+
+        ----- TX
+
+        let txE :: Either Text Tx
+            txE = mkTx (NE.fromList txIns) txOuts def
+        let tx = either (\e -> error $ "genTransaction impossible: couldn't create tx: " <> e)
+                        identity
+                        txE
+        let txId = hash tx
+        let taDistribution = TxDistribution $ NE.fromList $ replicate outputsN []
+        let toWitness :: TxIn -> Address -> TxInWitness
+            toWitness txIn = \case
+                PubKeyAddress stId _ ->
+                    let sk = resolveSecret stId
+                        txSig =
+                            sign SignTxIn sk TxSigData { txSigInput = txIn
+                                                       , txSigOutsHash = hash txOuts
+                                                       , txSigDistrHash = hash taDistribution }
+                    in PkWitness (toPublic sk) txSig
+                other -> error $
+                    sformat ("Found non-pubkey address: "%addressDetailedF) other
+        let taWitness =
+                V.fromList $ map (uncurry toWitness . second txOutAddress)
+                                 (txIns `zip` inputsResolved)
+        let txAux = TxAux { taTx = tx, .. }
```

Comment:
I don't like that you create tx fully by yourself. It's better to reuse something from `Pos.Client` or maybe add something new there.
Currently it doesn't account for tx fee checks and bootstrap era.
`Pos.Client` also doesn't account for it, but that's another problem. You can pretend that it takes these rules into account for now.